packages/server-wallet/src/onchain-service/onchain-service.ts

```diff
@@ -0,0 +1,173 @@
+} from './types';
+
+// FIXME: replace with
+// import {Wallet as ChannelWallet, WalletError as ChannelWalletError} from '@statechannels/server-wallet';
+import {Wallet as ChannelWallet, WalletError as ChannelWalletError} from '..';
+
+type Values<E> = E[keyof E];
+
+class OnchainServiceError extends ChannelWalletError {
+  readonly type = ChannelWalletError.errors.OnchainError;
+
+  // Reasons an error is thrown from the transaction submission
+  // service
+  static readonly reasons = {
+    notRegistered: 'Must call register channel',
+    noChannelWallet: 'Must attach channel wallet',
+  } as const;
+  constructor(
+    reason: Values<typeof OnchainServiceError.reasons>,
+    public readonly data: any = undefined
+  ) {
+    super(reason);
+  }
+}
+
+export class OnchainService implements OnchainServiceInterface {
+  private provider: providers.JsonRpcProvider;
+  private attempts: number;
+
+  // NOTE: V0 has an in-memory store, which will eventually be
+  // shifted to more permanent storage
+  // Holds latest channel event by type (i.e. { "fundsDeposited": {...} })
+  private store: Map<string, object> = new Map();
+
+  // Stores references to all contracts in memory
+  private assetHolders: Map<string, AssetHolderInformation & {contract: Contract}> = new Map();
+
+  // TODO: Remove reference to this object, and instead replace with
+  // outbox based communication. Right now used to call correct callbacks
+  private channelWallet: ChannelWallet | undefined = undefined;
+
+  // Handles the actual submission of the transactions to chain
+  private transactionSubmissionService: TransactionSubmissionServiceInterface;
+
+  constructor(
+    provider: string | providers.JsonRpcProvider,
+    transactionSubmissionService: TransactionSubmissionServiceInterface,
+    config: OnchainServiceConfiguration = {}
+  ) {
+    this.attempts = config.transactionAttempts || DEFAULT_MAX_TRANSACTION_ATTEMPTS;
+    if (this.attempts === 0) {
+      throw new Error(`Invalid number of attempts (${this.attempts})`);
+    }
+    this.provider =
+      typeof provider === 'string' ? new providers.JsonRpcProvider(provider) : provider;
+    this.transactionSubmissionService = transactionSubmissionService;
+  }
+
+  public attachChannelWallet(wallet: ChannelWallet): void {
+    this.channelWallet = wallet;
+  }
+
+  /**
+   * Adds a channel to watch events and submit transactions for
+   *
+   * @param channelId Unique channel identifier to register chain listeners for
+   * @param assetHolders Onchain addresses of the channel chain context
+   * @returns an empty promise that will resolve once the channel is registered
+   */
+  public registerChannel(
+    channelId: Bytes32,
+    assetHolders: AssetHolderInformation[]
+  ): Promise<void> {
+    if (!this.channelWallet) {
+      throw new OnchainServiceError(OnchainServiceError.reasons.noChannelWallet);
+    }
+
+    // If the channel has already been registered, return
+    if (this.store.has(channelId)) {
```

Comment:
What if the channel has already been registered, but for different asset holder addresses?

Reply:
I had not considered that possibility, and I rearranged some things to address this. As it stands:

- all registered channels are saved by channel id in the mapping with their events as the value (no longer any reference to asset holder info stored there)
- the asset holders are saved by asset holder address in the mapping, with an ethers Contract as the value

When a new channel is registered:
- IF it does not exist in the channels mapping, it is added (function does not return here anymore)
- IF the asset holder is not in the mapping, listeners for the contract events are registered